### PR TITLE
Make translate-c use intermediate AST

### PR DESCRIPTION
**Pull request by Vexu**
Will implement  #6710

### CODE REVIEW

**Comment by Sirius902** [comment]
Will this by chance address #3453?

**Comment by andrewrk** [comment]
Still working on this? Is there perhaps a way you can start the work in such a way that it doesn't have to be 100% completed before we can start utilizing it? That way we don't have a PR sitting around bit rotting.

**Comment by andrewrk** [comment]
Closing this old draft - the code is still available and linked from the respective issue description for anyone to reference it. When we tackle this again let's try to do it incrementally so that the whole rework/refactor does not have to get done before we can start merging into master branch.

**Comment by Vexu** [comment]
This currently doesn't affect `translate-c` and could be merged as is for aomeone to continue, ci failure was probably unrelated.

**Comment by andrewrk** [comment]
Ah I see, I will resurrect it and merge then!

**Comment by andrewrk** [comment]
This is going to make #7920 much easier to update translate-c because the only thing that will have change is the render() function.

**Comment by andrewrk** [comment]
Looks like the rebase got messed up

File: src/translate_c/ast.zig [hunk 1]
```diff
@@ -0,0 +1,267 @@
+const std = @import("std");
+const Type = @import("../type.zig").Type;
+
+pub const Node = struct {
```
Comment (andrewrk) [inline src/translate_c/ast.zig:4]: Looks like this is meant to be a union, yeah?

Comment (Vexu) [reply src/translate_c/ast.zig:4]: Yup, haven't yet gotten this to compile.

**Comment by andrewrk** [comment]
@Vexu whenever you feel this is ready to merge into `ast-memory-layout`, go for it

**Comment by ehaas** [comment]
There is one error happening on this branch that isn't happening with the old translate-c - duplicate non-translatable macros are getting emitted twice as `@compileError`.

```c
#define FOO 5=2
#define FOO 5=2
```
produces
```zig
pub const FOO = @compileError("unable to translate C expr: unexpected token .Equal"); // ddef.c:4:9
pub const FOO = @compileError("unable to translate C expr: unexpected token .Equal"); // ddef.c:5:9
```

Previously only one `FOO` was emitted. This happens to me on MacOS 11.2 when I `#include <stdlib.h>` due to duplicate definitions of `__API_AVAILABLE_PLATFORM_macCatalyst`.

**Comment by andrewrk** [comment]
One thing that will be interesting to test out on this branch, once we get it compiling again, is this benchmark. Here it is on master branch, on my machine:

#### test.c

```c
#include <windows.h>
```

(wipe global & local zig caches)

```
$ perf stat -e cache-misses -e instructions -e cycles ./zig translate-c test.c -target x86_64-windows-gnu -lc >windows_h.zig

 Performance counter stats for './zig translate-c test.c -target x86_64-windows-gnu -lc':

         6,632,603      cache-misses:u                                              
    40,151,754,610      instructions:u            #    1.39  insn per cycle         
    28,980,858,981      cycles:u                                                    

       6.669550308 seconds time elapsed

       6.539106000 seconds user
       0.132778000 seconds sys
```

**Pull request merged**